### PR TITLE
Move operatorOption.IPAMV4/V6CIDR and Mask out of operator to agent

### PR DESCRIPTION
**Pull request by soumynathan**
Related: #12186
Signed-off-by: Swaminathan Vasudevan <svasudevan@suse.com>

Please ensure your pull request adheres to the following guidelines:

### CODE REVIEW

**Comment by soumynathan** [comment]
@aanm @tklauser addressed review comments from 12186

File: operator/flags.go [hunk 1]
```diff
@@ -191,33 +191,9 @@ func init() {
...
-			option.IPAM, ipamOption.IPAMOperator,
-			option.EnableIPv6Name, "true"))
-	option.BindEnv(operatorOption.IPAMOperatorV6CIDR)
-
-	flags.Int(operatorOption.NodeCIDRMaskSizeIPv6, 112,
-		fmt.Sprintf("Mask size for each IPv6 podCIDR per node. Requires '%s=%s' and '%s=%s'",
-			option.IPAM, ipamOption.IPAMOperator,
-			option.EnableIPv6Name, "true"))
-	option.BindEnv(operatorOption.NodeCIDRMaskSizeIPv6)
-
```
Comment (aanm) [inline operator/flags.go]: This can't be removed otherwise `cilium-operator` won't receive the options.

File: daemon/cmd/daemon_main.go [hunk 2]
```diff
@@ -575,6 +575,26 @@ func init() {
...
+
+	flags.Int(option.ClusterPoolNodeCIDRMaskSizeIPv4, 24, fmt.Sprintf("Mask size for each IPv4 podCIDR per node. Requires '%s=%s' and '%s=%s'",
+		option.IPAM, ipamOption.IPAMOperator,
+		option.EnableIPv4Name, "true"))
+	option.BindEnv(option.ClusterPoolNodeCIDRMaskSizeIPv4)
+
+	flags.Int(option.ClusterPoolNodeCIDRMaskSizeIPv6, 112, fmt.Sprintf("Mask size for each IPv6 podCIDR per node. Requires '%s=%s' and '%s=%s'",
+		option.IPAM, ipamOption.IPAMOperator,
+		option.EnableIPv6Name, "true"))
+	option.BindEnv(option.ClusterPoolNodeCIDRMaskSizeIPv6)
```
Comment (aanm) [inline daemon/cmd/daemon_main.go]: This is not required, can be removed

File: daemon/cmd/daemon_main.go [hunk 3]
```diff
@@ -575,6 +575,26 @@ func init() {
...
 
+	flags.StringSlice(option.IPAMClusterPoolV4CIDR, []string{}, fmt.Sprintf("IPv4 CIDR Range for Pods in cluster. Requires '%s=%s' and '%s=%s'",
+		option.IPAM, ipamOption.IPAMOperator,
+		option.EnableIPv4Name, "true"))
+	option.BindEnv(option.IPAMClusterPoolV4CIDR)
+
+	flags.StringSlice(option.IPAMClusterPoolV6CIDR, []string{}, fmt.Sprintf("IPv6 CIDR Range for Pods in cluster. Requires '%s=%s' and '%s=%s'",
+		option.IPAM, ipamOption.IPAMOperator,
+		option.EnableIPv6Name, "true"))
+	option.BindEnv(option.IPAMClusterPoolV6CIDR)
```
Comment (aanm) [inline daemon/cmd/daemon_main.go]: These options should be marked as Hidden to not confuse the user.

**Review by aanm — CHANGES_REQUESTED** [review]

File: operator/flags.go (hunk 1 above)
Comment (soumynathan) [reply operator/flags.go]: Done

File: daemon/cmd/daemon_main.go (hunk 2 above)
Comment (soumynathan) [reply daemon/cmd/daemon_main.go]: Done.

File: daemon/cmd/daemon_main.go (hunk 3 above)
Comment (soumynathan) [reply daemon/cmd/daemon_main.go]: Done.

File: pkg/option/config.go [hunk 4]
```diff
@@ -2346,6 +2362,9 @@ func (c *DaemonConfig) Populate() {
 
 	c.populateDevices()
 
+	c.IPAMClusterPoolV4CIDR = viper.GetStringSlice(IPAMClusterPoolV4CIDR)
+	c.IPAMClusterPoolV6CIDR = viper.GetStringSlice(IPAMClusterPoolV6CIDR)
```
Comment (aanm) [inline pkg/option/config.go:2560]: a small note here, it's probably worth it make a small validation the the cidrs passed. If `net.ParseCIDR` would not return an error that would be sufficient.

**Review by aanm — CHANGES_REQUESTED** [review]

File: pkg/option/config.go (hunk 4 above)
Comment (soumynathan) [reply pkg/option/config.go:2560]: ok, will address your comments._

Comment (soumynathan) [reply pkg/option/config.go:2560]: Done.

File: pkg/option/config.go [hunk 5]
```diff
@@ -2022,6 +2038,36 @@ func (c *DaemonConfig) K8sAPIDiscoveryEnabled() bool {
 	return c.k8sEnableAPIDiscovery
 }
 
+func (c *DaemonConfig) validateIPAMClusterPoolV4CIDR() error {
+	if len(c.IPAMClusterPoolV4CIDR) != 0 {
```
Comment (aanm) [inline pkg/option/config.go]: You can remove this if statement, golang does not iterate over an empty size

File: pkg/option/config.go [hunk 6]
```diff
@@ -2022,6 +2038,36 @@ func (c *DaemonConfig) K8sAPIDiscoveryEnabled() bool {
...
+			if cidr == nil {
+				return fmt.Errorf("IPAMClusterPoolV4CDIR ParseCIDR returned nil")
+			}
+		}
+	}
+	return nil
+}
+
+func (c *DaemonConfig) validateIPAMClusterPoolV6CIDR() error {
+	if len(c.IPAMClusterPoolV6CIDR) != 0 {
```
Comment (aanm) [inline pkg/option/config.go]: You can remove this if statement, golang does not iterate over an empty size

**Review by aanm — CHANGES_REQUESTED** [review]

File: pkg/option/config.go (hunk 5 above)
Comment (soumynathan) [reply pkg/option/config.go]: Ok will do.

File: pkg/option/config.go (hunk 6 above)
Comment (soumynathan) [reply pkg/option/config.go]: Ok will do.

File: pkg/option/config.go (hunk 5 above)
Comment (soumynathan) [reply pkg/option/config.go]: Done.

File: pkg/option/config.go (hunk 6 above)
Comment (soumynathan) [reply pkg/option/config.go]: Done.

**Review by aanm — APPROVED** [review]

**Comment by aanm** [comment]
@soumynathan let's wait for the next Monday meeting to see if ti makes sense to merge this PR to fix https://github.com/cilium/cilium/issues/12186

**Comment by soumynathan** [comment]
> @soumynathan let's wait for the next Monday meeting to see if ti makes sense to merge this PR to fix #12186

@aanm thanks for the update. Sorry I was not able to attend the meeting today. Did you have any input from the meeting.

**Comment by soumynathan** [comment]
Rebased.

**Comment by aanm** [comment]
the failures seem related with the changes:
```
../pkg/option/config.go:2094:56: syntax error: unexpected error at end of statement
```

**Comment by soumynathan** [comment]
> the failures seem related with the changes:
> 
> ```
> ../pkg/option/config.go:2094:56: syntax error: unexpected error at end of statement
> ```

Will check this out.

**Review by aanm — APPROVED** [review]

**Comment by aanm** [comment]
test-me-please

File: pkg/option/config.go [hunk 7]
```diff
@@ -2429,6 +2471,10 @@ func (c *DaemonConfig) Populate() {
 
 	c.populateDevices()
 
+	c.IPAMClusterPoolV4CIDR = viper.GetStringSlice(IPAMClusterPoolV4CIDR)
+	c.validateIPAMClusterPoolV4CIDR()
+	c.IPAMClusterPoolV6CIDR = viper.GetStringSlice(IPAMClusterPoolV6CIDR)
+	c.validateIPAMClusterPoolV6CIDR()
```
Comment (joestringer) [inline pkg/option/config.go]: These lines are ignoring the returned error, but they should not.

**Review by joestringer — CHANGES_REQUESTED** [review]

**Comment by qmonnet** [comment]
Hi @soumynathan, did you have a chance to address Joe's feedback on this?

File: pkg/option/config.go (hunk 7 above)
Comment (soumynathan) [reply pkg/option/config.go]: Addressed review comments.

**Comment by soumynathan** [comment]
> Hi @soumynathan, did you have a chance to address Joe's feedback on this?

Just addressed review comments and rebased.

**Comment by qmonnet** [comment]
test-me-please

**Comment by aanm** [comment]
test-me-please

File: operator/option/config.go [hunk 8]
```diff
@@ -324,8 +308,6 @@ func (c *OperatorConfig) Populate() {
 	c.UnmanagedPodWatcherInterval = viper.GetInt(UnmanagedPodWatcherInterval)
 	c.NodeCIDRMaskSizeIPv4 = viper.GetInt(NodeCIDRMaskSizeIPv4)
 	c.NodeCIDRMaskSizeIPv6 = viper.GetInt(NodeCIDRMaskSizeIPv6)
-	c.IPAMOperatorV4CIDR = viper.GetStringSlice(IPAMOperatorV4CIDR)
-	c.IPAMOperatorV6CIDR = viper.GetStringSlice(IPAMOperatorV6CIDR)
```
Comment (joestringer) [inline operator/option/config.go:328]: For the operator, are these superseded by the calls in `DaemonConfig`? I'm not so familiar with this code but basically what I want to know is that we're not breaking this option for the operator. How do we establish this?

**Review by joestringer — CHANGES_REQUESTED** [review]
The title of the PR implies that the options are no longer in the operator, but the PR shows they still are. From previous feedback it seems like the goal is to have them in both. I think we just need to update the PR title?

One other piece of feedback below in an area I'm not so familiar with, I'm not sure if any code change is necessary there. Assuming these two things are addressed, this seems good to merge.

**Comment by aanm** [comment]
@soumynathan can you take one final look into this PR?

**Comment by mrostecki** [comment]
Converting to draft. I will overtake Swami's work and submit a new PR (also with #12230 included).

**Pull request closed without merging**